units/host_monitoring_parent_decommission.go

```diff
@@ -0,0 +1,142 @@
+)
+
+func init() {
+	registry.AddJobType(parentDecommissionJobName, func() amboy.Job {
+		return makeParentDecommissionJob()
+	})
+
+}
+
+type parentDecommissionJob struct {
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+	distroId string
+}
+
+func makeParentDecommissionJob() *parentDecommissionJob {
+	j := &parentDecommissionJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    parentDecommissionJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+
+	return j
+}
+
+func NewParentDecommissionJob(id string, distroId string) amboy.Job {
+	j := makeParentDecommissionJob()
+	j.distroId = distroId
+	j.SetID(fmt.Sprintf("%s.%s.%s", parentDecommissionJobName, j.distroId, id))
+	return j
+}
+
+// computeParentsToDecommission computes the number of hosts with containers to deco
+func computeHostsToDecommission(parentIds []string, distroId string) (int, error) {
+
+	numParents := len(parentIds)
+
+	// Count number of containers on the parents
+	containersQuery := db.Query(bson.M{
+		host.StatusKey:   bson.M{"$in": evergreen.UphostStatus},
+		host.ParentIDKey: bson.M{"$in": parentIds},
+	})
+	numContainers, err := host.Count(containersQuery)
+	if err != nil {
+		return 0, errors.Wrap(err, "Error counting containers on specified parents")
+	}
+
+	// Get maximum number of containers allowed on a parent with given distro
+	d, err := distro.FindOne(distro.ById(distroId))
+	if err != nil {
+		return 0, errors.Wrap(err, "Error getting max number of containers for distro")
+	}
+	maxContainersPerParent := d.MaxContainers
+
+	// Compute number of hosts to decommission based on excess capacity
+	numParentsToDeco := numParents - int(math.Ceil(float64(numContainers)/float64(maxContainersPerParent)))
+
+	// Sanity check
+	if numParentsToDeco < 0 {
+		numParentsToDeco = 0
+	}
+
+	return numParentsToDeco, nil
+}
+
+func (j *parentDecommissionJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	// Find hosts that will finish all container tasks soonest
+	parents, err := host.FindAllRunningParentsByDistro(j.distroId)
+	if err != nil {
+		j.AddError(err)
+		return
+	}
+
+	// Build list of parent Ids
+	var parentIds []string
+	for _, parent := range parents {
+		parentIds = append(parentIds, parent.Id)
+	}
+
+	numParentsToDeco, err := computeHostsToDecommission(parentIds, j.distroId)
+	if err != nil {
+		j.AddError(err)
+		return
+	}
+
+	// Get desired number parents with nearest LastContainerFinishTime
+	parentsToDeco := parents[:numParentsToDeco]
+
+	for _, h := range parentsToDeco {
+
+		// Decommission each container on parent host
+		containersToDeco, err := h.GetContainers()
+		if err != nil {
+			j.AddError(err)
+			return
+		}
+
+		for _, c := range containersToDeco {
+			if c.Status == evergreen.HostRunning {
+				err := c.SetDecommissioned(evergreen.User, "")
+				j.AddError(err)
+			}
+		}
+
+		// Decommission parent only if all containers have terminated
+		idle, err := h.IsIdleParent()
```

Comment:
Does this run into this problem?
1. We mark a parent to be deco'd and deco its containers
2. In the 10 minutes between when its containers are deco'd and the parent is reaped, something happens such that the parent no longer sorts to the part of the FindAllRunningParentsByDistro query that will get deco'd

I remember discussing this but don't remember if we concluded anything

Reply:
I don't think this is a problem -- it's fine if we intend to deco a parent host (thereby deco-ing all of its containers) and then no longer wish to decommission that parent before all of its containers terminates. This gives us some built-in flexibility for cases where demand changes and we don't want to go through with terminating parents that are no longer unnecessary.